### ISSUE TITLE
patch names blank in Evolver desktop dump

### ISSUE THREAD

**Issue by ergodicbreak** [christofmuc/KnobKraft-orm#115]
Hello, first thanks very much for KnobKraft, it looks nice and works smoothly! I cloned and built on Manjaro Linux. 

I have an issue I think related to the alpha status of the Evolver adaptation, where all the patch names are blank (and editing a patch doesn't add the name -- it stays blank). 

From the Python adaptation file TODO list it seems this is an issue with the Evolver specifically. I am willing to work on the adaptation to improve it, but since this is my first time with KnobKraft (and I am complete beginner with synths and MIDI) I would appreciate any advice/direction on how to proceed. 

Screenshot below, 

![knobkraft_imported](https://user-images.githubusercontent.com/3825386/137446483-c3a21fb8-8293-4eae-a9dd-4922befa28fe.png)

**Comment by ergodicbreak** [comment]
I posted a message on Sequential, 

https://forum.sequential.com/index.php/topic,5757.0.html

I have a feeling the program name on the Desktop is only there for compatibility and you can't change it, which means there'll have to be some solution to associate names in KnobKraft with some kind of patch fingerprint.

**Comment by christofmuc** [comment]
Haha, that looks like quite a mess in Evolver firmwares!

So I understand:

  * To keep backwards compability of the newer Evolvers with the original device (which is ?), they decided to add the program name to the existing sysex message, but to add a second message with just the name.
  * Some Evolvers (not the Desktop) send the second message automatically when a program dump is request. For the Desktop, you need to send the explicit Name Request in addition to the Program Dump request
  * I am still not clear whether the Edit Buffer Dump also has a name? 

KnobKraft functions as follows: 

  * The database stores a byte vector for each patch. This could be the original MidiMessage of the patch, or multiple messages concatenated, or some processed form. It additionally stores a name in a separate data field to allow naming of patches even for synths that do not have name storage
  * The database will not allow duplicates to be imported, which is based on fingerprinting the byte vector. Normally, you want to blank out the name before comparison, because otherwise patches that are identical except the name will create a new database entry. This functionality is enabled for Python adaptations with the calculateFingerprint method, which you can reimplement.
  * The timing of some devices is fidgety. As you are allowed to return multiple MIDI messages as a response to requestProgramDump() from Python, you could just append the Request Name Dump message - ideally, the Evolver will buffer and send out both replies. Even if the name dump is not processed right now, you should see it incoming in the MIDI log of the Orm. 
  
Also, sending a patch to the synth can create multiple messages (patchToProgramDumpSysex returns a vector of MIDI messages), so we are already good to send out multiple messages to the synth.

This leaves us with the one problem that the current simple parsing logic only wants one message per patch.

Today, I have looked at the code from January this year,  and it might not be too far off to bring it into production. That would solve this problem (and many many more for other synths). Have a look at https://github.com/christofmuc/KnobKraft-orm/blob/banksFromDataTypes/adaptions/GenericProgramDumpCapability.h at let me know if that does make any sense.

A more complex synth that uses 7 messages per patch (typcial for Yamaha) is the refaceDX, for which I made a prototype implementation here in Python (so far, it had to be C++): https://github.com/christofmuc/KnobKraft-orm/blob/banksFromDataTypes/adaptions/YamahaRefaceDX.py

My feeling is that work should be completed before we decide to do any other shortcut for the Evolver - I currently can't think of something simpler, but that is maybe because I already was way into this general stuff ;-)
  *

**Comment by ergodicbreak** [comment]
> My feeling is that work should be completed before we decide to do any other shortcut for the Evolver - I currently can't think of something simpler, but that is maybe because I already was way into this general stuff

I agree this makes the most sense. 

> This leaves us with the one problem that the current simple parsing logic only wants one message per patch.

Are you saying the Python adaptation of the Yamaha DX is using new functionality? I'm wondering based on this, 

https://github.com/christofmuc/KnobKraft-orm/blob/ccbd1dd73540cb885a8a6d9f1a014e038bbc81ca/adaptions/YamahaRefaceDX.py#L47

Where the code comment is `# Actually, those are multiple messages`; does this mean we could send two sysex messages to the Evolver Desktop like @composerjk  did in his last comment and parse the returning two messages like this `nameFromDump` does?

**Comment by ergodicbreak** [comment]
> Maybe we could just specify a "number of messages per patch" and then treat them as one array input into the isProgramDump() / nameFromDump() messages? That could be simpler than the generic stream load capability I have in the branch mentioned above

Yes a simple declarative solution sounds good, unless there are indeed some synths with not well defined responses?

**Comment by ergodicbreak** [comment]
Sorry to be a bother, but I'm getting the same error as before. Just in case I'm missing something (which is likely!) here are the steps I took in a clean directory:

# git clone --recurse-submodules https://github.com/christofmuc/KnobKraft-orm
# cd KnobKraft-orm
# git checkout multiMessageEditBuffers
# LDFLAGS="--disable-lto" cmake -S . -B builds && cmake --build builds

The error was

> -- Exporting juceaide
> CMake Error at CMakeLists.txt:194 (add_subdirectory):
>   The source directory
> 
>     /home/george/work/tools/KnobKraft-orm/third_party/json
> 
>   does not contain a CMakeLists.txt file.
> 
> 
> CMake Error at CMakeLists.txt:195 (add_subdirectory):
>   The source directory
> 
>     /home/george/work/tools/KnobKraft-orm/third_party/json-schema-validator
> 
>   does not contain a CMakeLists.txt file.
> 
> Log here https://gist.github.com/ergodicbreak/fca31b2405571159c12b45c9811b49ec
> ##vso[task.setvariable variable=ormVersionNumber]1.14.0
> -- Configuring incomplete, errors occurred!

https://gist.github.com/ergodicbreak/fca31b2405571159c12b45c9811b49ec

**Comment by christofmuc** [comment]
@ergodicbreak Thanks for the update! My mistake to not add the submodules in the first place, but yes, pulling the latest version and then running the proper git vodoo (probably git submodule update --init --recursive) should do the job. Let me know how you progress with the Evolver!
